core/src/main/java/kafka/server/share/DelayedShareFetchKey.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.server.share;
+
+import kafka.server.DelayedOperationKey;
+
+import org.apache.kafka.common.TopicIdPartition;
+
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A key for delayed operations that fetch data for share consumers.
+ */
+public class DelayedShareFetchKey implements DelayedOperationKey {
+    private final Set<TopicIdPartition> topicIdPartitions;
```

Comment:
Each entry in the new purgatory should correspond to a share fetch, but the keys should be more granular. This is really the power of request purgatories. The keys represent dimensions of the entry which can cause it to be completed. 

For fetch requests, any partition in the request can cause the request to be completed via MinBytes. So, we create a DelayedFetchRequest with keys for each partition. https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ReplicaManager.scala#L1587-L1601

For share fetch, we'll do something similar. For example:

* SFR1 arrives (SP1, SP2, SP3), no locks available so we create DSF1
* Purgatory now contains one entry DSF1 with three keys SPK1, SPK2, SPK3 -- one key per share partition
* SFR2 arrives with acks for SP1 from another member, we need to call tryComplete for SPK1
* The completion of SPK1 may complete SFR1 if MinBytes is hit

Reply:
Yes, I agree that it would be useful to think through the key of DelayedShareOperation. Intuitive, two things could trigger the evaluation of a DelayedShareOperation: (1) new data on the partition; (2) new acknowledgement from a member. If (1) occurs, it affects every share fetch request on that partition. If (2) occurs, it affects every share fetch request on that partition in the same group. To cover both case, having a key with partition + groupId seems enough. Not sure why we need memberId as part of the key.

Reply:
Hi @mumrah @junrao , I have updated DelayedShareFetchKey to have two attributes - 
1. TopicIdPartition
2. GroupId